blog/_src/posts/2018-10-05-defining-local-bindings-in-turnstile-languages.md

```diff
@@ -0,0 +1,623 @@
+    Title: Defining Local Bindings in Turnstile Languages
+    Date: 2018-10-05T15:05:17
+    Tags: by Sam Caldwell
+
+In [Racket][2], programmers can create powerful abstractions by bundling
+together a family of values, functions, and syntax extensions in the form of a
+new language. These languages are typically untyped. [Turnstile][1] is a new
+Racket {library,language} for creating typed languages by
+integrating type checking with Racket's existing tools for
+describing languages. The technique is described by fellow PRL'ers
+in the paper [*Type Systems as Macros*][3].
+
+Racket encourages language developers to take full advantage of [linguistic
+reuse][21] by defining new language forms in terms of existing constructs.
+Unsurprisingly, language extensions often retain some of the Racket-y flavor
+from the underlying constructs. Implementors save time and energy while users of
+the language benefit from the familiarity they already have with the Racket
+ecosystem.
+
+Unfortunately, Turnstile does not lend itself to expressing one of Racket's most
+ubiquitous idioms: naming local bindings with `define`. Early experience reports
+from Turnstile, including my own, suggest that language implementors very much
+desire to include `define`-like binding forms in their languages.
+
+This blog post provides a brief overview of what Turnstile is and how it works,
+an introduction to defining typed language forms, and how to equip these
+languages with a `define` binding form.
+
+<!-- more -->
+
+The code for this blog post can be found [in this gist][14]. To run it, you will
+need the Turnstile package, which can be installed with `raco pkg install
+turnstile`.
+
+## Turnstile: Typechecking Intertwined with Elaboration
+
+Turnstile provides a convenient way of defining syntax transformations that also
+perform typechecking. Since processing the syntax of a form typically involves
+some amount of analysis, such as for error checking, it is a natural place to
+put the logic for typechecking. With forms defined as such, macro expanding a
+program determines both a type and an elaborated term in the target language.
+
+While macro expansion proceeds outside-in, type information typically flows up
+from the leaves during checking. To reconcile the two directions, Turnstile
```

Comment:
"up from the leaves" surprised me, maybe instead:
- "flows inside-out during checking"
- "up from the leaves of the AST during checking"